### ISSUE TITLE
Segmentation fault in ConvNet built with a slightly unusual structure

### ISSUE THREAD

**Issue by stablum** [Lasagne/Lasagne#598]
Hi,

I am obtaining a `segmentation fault`, that is something that I think should never-ever happen.
The segfault is triggered by the training function.
The network has this structure:
```
InputLayer (100,3,128,64)
Conv2DLayer
	filter_size (3, 3)
	nonlinearity: tanh
	untie_biases: True
	pad: same
Conv2DLayer
	filter_size (5, 5)
	nonlinearity: softplus
	untie_biases: False
	pad: (0, 0)
MaxPool2DLayer
	stride: (10, 10)
	pool_size (6, 6)
MaxPool2DLayer
	stride: (4, 4)
	pool_size (6, 6)
DropoutLayer
	p: 0.759374613886
DropoutLayer
	p: 0.0429295503475
DenseLayer
	nonlinearity: softmax
	num_units: 549
DenseLayer
	nonlinearity: linear
	num_units: 2
```
any ideas on why this could happen?

**Comment by stablum** [comment]
another structure that segfaults:
InputLayer
Conv2DLayer
	filter_size (8, 8)
	nonlinearity: LeakyRectify
	untie_biases: False
	pad: (0, 0)
MaxPool2DLayer
	stride: (11, 11)
	pool_size (6, 6)
Conv2DLayer
	filter_size (3, 3)
	nonlinearity: LeakyRectify
	untie_biases: False
	pad: (0, 0)
DropoutLayer
	p: 0.0775675637794
DropoutLayer
	p: 0.59428341011
DenseLayer
	nonlinearity: LeakyRectify
	num_units: 927
DenseLayer
	nonlinearity: sigmoid
	num_units: 621
DropoutLayer
	p: 0.730677121429
DenseLayer
	nonlinearity: softmax
	num_units: 2

**Comment by stablum** [comment]
Here is another one (notice I added the input/output shapes):
```
InputLayer  (100, 3, 128, 64)
Conv2DLayer (100, 3, 128, 64) --> (100, 449, 127, 63)
	filter_size (2, 2)
	nonlinearity: tanh
	untie_biases: True
	pad: (0, 0)
MaxPool2DLayer (100, 449, 127, 63) --> (100, 449, 124, 60)
	stride: (1, 1)
	pool_size (4, 4)
MaxPool2DLayer (100, 449, 124, 60) --> (100, 449, 61, 29)
	stride: (2, 2)
	pool_size (4, 4)
MaxPool2DLayer (100, 449, 61, 29) --> (100, 449, 20, 9)
	stride: (3, 3)
	pool_size (4, 4)
MaxPool2DLayer (100, 449, 20, 9) --> (100, 449, 7, 3)
	stride: (3, 3)
	pool_size (2, 2)
DropoutLayer (100, 449, 7, 3) --> (100, 449, 7, 3)
	p: 0.128847397392
DenseLayer (100, 449, 7, 3) --> (100, 2033)
	nonlinearity: linear
	num_units: 2033
DropoutLayer (100, 2033) --> (100, 2033)
	p: 0.0157800160789
DropoutLayer (100, 2033) --> (100, 2033)
	p: 0.76798092819
DenseLayer (100, 2033) --> (100, 2)
	nonlinearity: softmax
	num_units: 2
```
Is it possible that segfaults when too much memory is being used?

**Comment by nouiz** [comment]
Can you give a full script that reproduce this? Debugging segfault is hard.
Normally, not enough memory should not cause segfault except if you changed
python recursion limit.

**Comment by f0k** [comment]
Also can you update to the latest (bleeding-edge) versions of Theano and Lasagne? If anything, it's a Theano bug, and it may have been fixed in the meantime. Update instructions: http://lasagne.readthedocs.org/en/latest/user/installation.html#bleeding-edge-version

**Comment by stablum** [comment]
The installation of Theano and Lasagne is quite fresh; these are the version numbers:
```
>>> import theano
t>>> theano.__version__
'0.7.0.dev-54186290a97186b9c6b76317e007844529a352f4'
>>> import lasagne
>>> lasagne.__version__
'0.2.dev1'
```

**Comment by stablum** [comment]
unfortunately giving the code is not so useful since it's a complex system configured by a genetic algorithm that "invents" a new structure every time.

**Comment by stablum** [comment]
Here is another network that causes the segfault:
```
InputLayer  (100, 3, 128, 64)
Conv2DLayer (100, 3, 128, 64) --> (100, 191, 128, 64)
	filter_size (7, 7)
	nonlinearity: tanh
	untie_biases: True
	pad: same
MaxPool2DLayer (100, 191, 128, 64) --> (100, 191, 21, 10)
	stride: (6, 6)
	pool_size (5, 5)
Conv2DLayer (100, 191, 21, 10) --> (100, 8, 21, 10)
	filter_size (7, 7)
	nonlinearity: linear
	untie_biases: False
	pad: same
DenseLayer (100, 8, 21, 10) --> (100, 148)
	nonlinearity: softplus
	num_units: 148
DenseLayer (100, 148) --> (100, 427)
	nonlinearity: softplus
	num_units: 427
DenseLayer (100, 427) --> (100, 310)
	nonlinearity: tanh
	num_units: 310
DropoutLayer (100, 310) --> (100, 310)
	p: 0.924510344398
DenseLayer (100, 310) --> (100, 2)
	nonlinearity: tanh
	num_units: 2
starting batch training...
Segmentation fault (core dumped)
```
It's using `squared_error` loss function and `adam` updates, with `learning_rate=1.34837933921`.

**Comment by f0k** [comment]
> unfortunately giving the code is not so useful since it's a complex system configured by a genetic algorithm that "invents" a new structure every time.

Sure, but you could write a script that instantiates one of the networks you posted, in exactly the way it would have been instantiated by your genetic algorithm, and make whatever call is necessary to trigger the segfault. It's impossible to debug this without a ready-to-run script that triggers the segfault somewhat reliably! Maybe it's not even related to the network architecture?
You can also try to run this with `gdb` so it will give you a backtrace at the segfault. It takes some extra steps to get `gdb` to backtrace into your Python code, but it's possible. But still, trying to craft a single script that triggers the segfault (even if just in 10% of the cases) would be the first step!

**Comment by nouiz** [comment]
If you want to try gdb, here is some instruction:

https://github.com/Theano/Theano/pull/3895/files

Sadly, we have problems that the online doc not regenerated for now.

**Comment by stablum** [comment]
The problem was likely caused by a lack of memory. Too bad python is not communicating that properly and just segfaults.

**Comment by nouiz** [comment]
Normally, there is an error message. It could be in Theano too. Some older
version of numpy had problem like that too.

If you want, you could relatively easily investigate to find where the
problem come from by starting the process in gdb. It would tell us which
software the problem is in and if it is in Theano, we can look at it.

**Comment by f0k** [comment]
Closing because there was no more feedback from the OP. Please feel free to comment again if you can still reproduce the problem, it would be nice to fix the origin of the segfault if at all possible.
